File: src/main/java/architecture/goldenboughs_lib/init/LibMobEffects.java
package architecture.goldenboughs_lib.init;

import architecture.goldenboughs_lib.api.LcDamageType;
import architecture.goldenboughs_lib.common.mobeffect.MobEffectExpand;
import architecture.goldenboughs_lib.core.Lib;
import architecture.goldenboughs_lib.datagen.i18n.LibZhCn;
import net.minecraft.core.Holder;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.effect.MobEffectCategory;
import net.neoforged.neoforge.registries.DeferredHolder;
import net.neoforged.neoforge.registries.DeferredRegister;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;

public final class LibMobEffects {
	public static final DeferredRegister<MobEffect> REGISTRY = Lib.modRegister(BuiltInRegistries.MOB_EFFECT);

	public static final Holder<MobEffect> PHYSIC_ABSORPTION_SHIELD = register("physic_absorption_shield", "物理吸收护盾",
		MobEffectExpand::new, MobEffectCategory.BENEFICIAL, LcDamageType.PHYSICS.getColourValue());
	public static final Holder<MobEffect> SPIRIT_ABSORPTION_SHIELD = register("spirit_absorption_shield", "精神吸收护盾",
		MobEffectExpand::new, MobEffectCategory.BENEFICIAL, LcDamageType.SPIRIT.getColourValue());
	public static final Holder<MobEffect> EROSION_ABSORPTION_SHIELD = register("erosion_absorption_shield", "侵蚀吸收护盾",
		MobEffectExpand::new, MobEffectCategory.BENEFICIAL, LcDamageType.EROSION.getColourValue());
	public static final Holder<MobEffect> SOUL_ABSORPTION_SHIELD = register("soul_absorption_shield", "灵魂吸收护盾",
		MobEffectExpand::new, MobEffectCategory.BENEFICIAL, LcDamageType.SPIRIT.getColourValue());

	private static <T extends MobEffect> DeferredHolder<MobEffect, T> register(String name, String zhCnText, BiFunction<MobEffectCategory, Integer, T> biFunction, MobEffectCategory category, int color) {
		return register(name, zhCnText, () -> biFunction.apply(category, color));
	}

	private static <T extends MobEffect> DeferredHolder<MobEffect, T> register(String name, String zhCnText, Supplier<T> supplier) {
		DeferredHolder<MobEffect, T> holder = REGISTRY.register(name, supplier);
		LibZhCn.addI18nMobEffectText(zhCnText, holder);
		return holder;
	}

	private static <T extends MobEffect> DeferredHolder<MobEffect, T> register(String name, String zhCnText, Supplier<T> supplier, Function<T, MobEffect> function) {
		return register(name, zhCnText, () -> {
			T apply = supplier.get();
			function.apply(apply);
			return apply;
		});
	}

	private static <T extends MobEffect> DeferredHolder<MobEffect, T> register(String name, String zhCnText, Supplier<T> supplier, BiFunction<T, ResourceLocation, MobEffect> function) {
		return register(name, zhCnText, () -> {
			T apply = supplier.get();
			function.apply(apply, Lib.modRl(name));
			return apply;
		});
	}

	private static <T extends MobEffect> DeferredHolder<MobEffect, T> register(String name, String zhCnText, BiFunction<MobEffectCategory, Integer, T> biFunction, MobEffectCategory category, int color, Function<T, MobEffect> function) {
		return register(name, zhCnText, () -> {
			T apply = biFunction.apply(category, color);
			function.apply(apply);
			return apply;
		});
	}

	private static <T extends MobEffect> DeferredHolder<MobEffect, T> register(String name, String zhCnText, BiFunction<MobEffectCategory, Integer, T> biFunction, MobEffectCategory category, int color, BiFunction<T, ResourceLocation, MobEffect> function) {
		return register(name, zhCnText, () -> {
			T apply = biFunction.apply(category, color);
			function.apply(apply, Lib.modRl(name));
			return apply;
		});
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/init/LibMobEffects.kt
package architecture.goldenboughs_lib.init

import architecture.goldenboughs_lib.core.LibConstants.modRegister
import architecture.goldenboughs_lib.core.LibConstants.modRl
import architecture.goldenboughs_lib.datagen.i18n.LibZhCn
import architecture.goldenboughs_lib.module.lc_damage.api.LcDamageType
import architecture.goldenboughs_lib.module.shield.mobeffect.MobEffectExpand
import net.minecraft.core.Holder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.effect.MobEffectCategory
import net.neoforged.neoforge.registries.DeferredHolder
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.BiFunction
import java.util.function.Function
import java.util.function.Supplier

object LibMobEffects {
	@JvmField
	val REGISTRY: DeferredRegister<MobEffect> = modRegister(BuiltInRegistries.MOB_EFFECT)

	@JvmField
	val PHYSIC_ABSORPTION_SHIELD: Holder<MobEffect> = register(
		"physic_absorption_shield",
		"物理吸收护盾",
		::MobEffectExpand,
		MobEffectCategory.BENEFICIAL,
		LcDamageType.PHYSICS.colourValue
	)

	@JvmField
	val SPIRIT_ABSORPTION_SHIELD: Holder<MobEffect> = register(
		"spirit_absorption_shield",
		"精神吸收护盾",
		::MobEffectExpand,
		MobEffectCategory.BENEFICIAL,
		LcDamageType.SPIRIT.colourValue
	)

	@JvmField
	val EROSION_ABSORPTION_SHIELD: Holder<MobEffect> = register(
		"erosion_absorption_shield",
		"侵蚀吸收护盾",
		::MobEffectExpand,
		MobEffectCategory.BENEFICIAL,
		LcDamageType.EROSION.colourValue
	)

	@JvmField
	val SOUL_ABSORPTION_SHIELD: Holder<MobEffect> = register(
		"soul_absorption_shield",
		"灵魂吸收护盾",
		::MobEffectExpand,
		MobEffectCategory.BENEFICIAL,
		LcDamageType.SPIRIT.colourValue
	)

	private fun <T : MobEffect> register(
		name: String,
		zhCnText: String,
		biFunction: BiFunction<MobEffectCategory, Int, T>,
		category: MobEffectCategory,
		color: Int
	): DeferredHolder<MobEffect, T> {
		return register(name, zhCnText) { biFunction.apply(category, color) }
	}

	private fun <T : MobEffect> register(
		name: String, zhCnText: String, supplier: Supplier<T>
	): DeferredHolder<MobEffect, T> {
		val holder = REGISTRY.register(name, supplier)
		LibZhCn.addI18nMobEffectText(zhCnText, holder)
		return holder
	}

	private fun <T : MobEffect> register(
		name: String, zhCnText: String, supplier: Supplier<T>, function: Function<T, MobEffect>
	): DeferredHolder<MobEffect, T> {
		return register(name, zhCnText) {
			val apply = supplier.get()
			function.apply(apply)
			apply
		}
	}

	private fun <T : MobEffect> register(
		name: String, zhCnText: String, supplier: Supplier<T>, function: BiFunction<T, ResourceLocation, MobEffect>
	): DeferredHolder<MobEffect, T> {
		return register(name, zhCnText) {
			val apply = supplier.get()
			function.apply(apply, modRl(name))
			apply
		}
	}

	private fun <T : MobEffect> register(
		name: String,
		zhCnText: String,
		biFunction: BiFunction<MobEffectCategory, Int, T>,
		category: MobEffectCategory,
		color: Int,
		function: Function<T, MobEffect>
	): DeferredHolder<MobEffect, T> {
		return register(name, zhCnText) {
			val apply = biFunction.apply(category, color)
			function.apply(apply)
			apply
		}
	}

	private fun <T : MobEffect> register(
		name: String,
		zhCnText: String,
		biFunction: BiFunction<MobEffectCategory, Int, T>,
		category: MobEffectCategory,
		color: Int,
		function: BiFunction<T, ResourceLocation, MobEffect>
	): DeferredHolder<MobEffect, T> {
		return register(name, zhCnText) {
			val apply = biFunction.apply(category, color)
			function.apply(apply, modRl(name))
			apply
		}
	}
}
